src/backend/distributed/utils/metadata_cache.c

```diff
@@ -1217,13 +1228,15 @@ BuildCachedShardList(DistTableCacheEntry *cacheEntry)
 
 		cacheEntry->arrayOfPlacementArrays[shardIndex] = placementArray;
 		cacheEntry->arrayOfPlacementArrayLengths[shardIndex] = numberOfPlacements;
+		cacheEntry->shardIntervalArrayLength++;
 
 		/* store the shard index in the ShardInterval */
 		shardInterval->shardIndex = shardIndex;
+		RESUME_INTERRUPTS();
+
+		CHECK_FOR_INTERRUPTS();
```

Comment:
I'm wondering whether we should just put the HOLD_INTERRUPTS around the whole loop

Reply:
I think the code restructuring makes sense regardless. We may still have other errors that mess up the cache entry (e.g. out of memory), though we're probably not completely protected against those yet if e.g. what if an error occurs in `BuildShardPlacementList`? Should we increment the `shardIntervalArrayLength` immediately after adding an item to the hash such that we always remove the shard from the hash later, even if an error occurred immediately after adding it?

Reply:
Got it. Thanks for the explanation.

Looking at the code, I think we should either move `HOLD_INTERRUPTS` to `LookupDistTableCacheEntry()`, or `CHECK_FOR_INTERRUPTS()` in each iteration of this loop.

If we just do this loop in a `HOLD_INTERRUPTS`, the cancellation can happen much later if shard count is high, and we aren't actually creating anything that we will reuse later, since `DistTableCacheEntry->isValid` will stay false and we will rebuild the cache data again later.

Because if the cancellation is happening because of `statement_timeout`, I'd vote for moving `HOLD_INTERRUPTS` to `LookupDistTableCacheEntry()` so we don't get timeouts over and over again.

What do you think?